src/sentry/grouping/ingest/seer.py

```diff
@@ -0,0 +1,95 @@
+import logging
+from dataclasses import asdict
+
+from sentry import features
+from sentry.api.endpoints.group_similar_issues_embeddings import get_stacktrace_string
+from sentry.constants import PLACEHOLDER_EVENT_TITLES
+from sentry.eventstore.models import Event
+from sentry.grouping.grouping_info import get_grouping_info_from_variants
+from sentry.grouping.result import CalculatedHashes
+from sentry.models.group import Group
+from sentry.models.project import Project
+from sentry.seer.utils import (
+    SeerSimilarIssuesMetadata,
+    SimilarIssuesEmbeddingsRequest,
+    get_similarity_data_from_seer,
+)
+from sentry.utils.safe import get_path
+
+logger = logging.getLogger("sentry.events.grouping")
+
+
+def should_call_seer_for_grouping(event: Event, project: Project) -> bool:
+    """
+    Use event content, feature flags, rate limits, killswitches, seer health, etc. to determine
+    whether a call to Seer should be made.
+    """
+    # TODO: Implement rate limits, kill switches, other flags, etc
+    # TODO: Return False if the event has a custom fingerprint (check for both client- and server-side fingerprints)
+
+    if (
+        event.title in PLACEHOLDER_EVENT_TITLES
+        and not get_path(event.data, "exception", "values", -1, "stacktrace", "frames")
+        and not get_path(event.data, "threads", "values", -1, "stacktrace", "frames")
+    ):
+        return False
+
+    return features.has("projects:similarity-embeddings-metadata", project) or features.has(
+        "projects:similarity-embeddings-grouping", project
+    )
+
+
+def get_seer_similar_issues(
+    event: Event,
+    primary_hashes: CalculatedHashes,
+    num_neighbors: int = 1,
+) -> tuple[
+    dict[
+        str, str | list[dict[str, float | bool | int | str]]
+    ],  # a SeerSimilarIssuesMetadata instance, dictified
+    Group | None,
+]:
+    """
+    Ask Seer for the given event's nearest neighbor(s) and return the seer response data, sorted
+    with the best matches first, along with the group Seer decided the event should go in, if any,
+    or None if no neighbor was near enough.
+
+    Will also return `None` for the neighboring group if the `projects:similarity-embeddings-grouping`
+    feature flag is off.
+    """
+
+    # TODO: Once we get rid of hierarchical hashing, we'll be able to make `variants` required in
+    # `CalculatedHashes`, meaning we can remove this check. (See note in `CalculatedHashes` class definition.)
+    if primary_hashes.variants is None:
+        raise Exception("Primary hashes missing variants data")
```

Comment:
can we raise a more specific exception? can also just define your own exception that inherits from `Exception`

Reply:
The thing is, this never happens. There are other scenarios where `variants` can be `None`, but this isn't one of them. We need this bit mostly to make mypy happy, since those other scenarios force an `| None` into the type. I can make this clearer in the comment.

As the comment already says, though, all of that is only true (and therefore this check only exists) until we kill the hierarchical grouping code, which is why I didn't go out of my way to make it a super useful error.